### DISCUSSION TITLE
Tips received for Dogecoin Core contributors

### DISCUSSION THREAD

**Discussion by patricklodder** [dogecoin/dogecoin#3073] [General]
As you may have realized, I often receive tips for work done on this repository, especially whenever we do a release. These tips however aren’t mine alone: they belong to everyone that did the work to get to the release, because our users often don't know who did what. Therefore, I have saved those tips up and “vaulted” them in a time-locked p2sh address. Since I [announced the vault](https://www.reddit.com/user/patricklodder/comments/v9lvlb/accounting_for_dogecoin_core_development_tips/), a lot of shibes have been tipping the vault directly. This is awesome. 

Unfortunately I cannot deposit those tips into the tipjar we’ve maintained since December 2014, because the published payout policy was violated and this has yet to be resolved. I do not have any idea if or when it will be resolved, and I do not want any contributors to fall victim to the endless stream of flip-flopping and delaying that has manifested itself so far. This means that I now intend to distribute tips to contributors separately and independently from what happens with the historical tipjar. Note that there are still tips incoming for the 1.14.6 release into the vault to me personally, so I think it makes sense to wait just a little longer with payouts. The address of the vault is [9rbpxxkjB9uZbcZzZCPibnoiotGkPoGFEJ](https://dogeblocks.com/address/9rbpxxkjB9uZbcZzZCPibnoiotGkPoGFEJ).

Here’s what I’m intending to do to get to a fair and inclusive payout:

1. Each contribution to a merged pull request will be weighted and assigned a "weight": 1 for tiny work, 5 for a normal pull request, 20 for a big change. I will personally weigh this and publish my results on GitHub. Reviewers will be included in the rewards, as long as there was - to my judgement - non-trivial work done for the review (i.e. hitting the approve button just because you can is trivial, code reviewing 2 lines is trivial, building and testing a PR is non-trivial)
2. Each contributor will have a sum of “weight” over the work done. This will be defining the payout height for each individual according to: `total received / total weight * individual weight` 
3. In parallel to this, I need to collect payout addresses. Each contributor that I can reach will receive an email. If you don’t want to expose your email address, simply put a Dogecoin address on your GitHub profile and I’ll copy it from there.
4. On August 21st - which is exactly one month after release, I want to make the share of tips received available to each contributor, distributing the total amount I received until that date.

If you have any questions, comments or suggestions, just leave them here.

_To all code-contributing shibes: thank you for being awesome and generous with your time and effort._
_To all tipping shibes: thank you for being awesome and generous with your DOGE._

**Comment by MishaBoar** [comment]
The criteria seem to be fine to me, So I am OK with this.

Those doing maintainer's work (and there was a crazy amount of that to coordinate so much talent and so many contributors new to the process) should be rewarded again, of course, too.

I hope the old tipjar issue is resolved as soon as possible. As a Dogecoin's user, it pains and saddens me.

**Reply by patricklodder** [reply]
> Those doing maintainer's work (and there as a crazy amount of that to coordinate so much talent and so many contributors new to the process) should be rewarded again, of course, too.

Maintainer work is done through PRs and reviewing is a large part of that work, so that is automatically included in the above.

I received a tiny note about security researchers elsewhere and I think the point made is actually a good one: they also do important work! Therefore, since security researchers do similar work to review, but _before_ there is a PR, I propose to include these contributors to get a share of the tips the same as reviewers, for everything that made it in.

**Reply by MishaBoar** [reply]
I agree!

**Comment by alamshafil** [comment]
I agree with the payout since it is fair to all shibes who contributed. 

The fact you are publishing your results increases transparency and I appreciate it. This gives the community the ability to comment on the results. 

Including the non-trival reviewers is also important because reviewing and testing code is not easy. 

As already stated. I feel like the maintainers should be included in the payout since they were instrumental in helping shibes contribute correctly and getting their contributions merged. 

Maintaining a whole cryptocurrency at this level is extremely time consuming and full of hard work. I appreciate the effort you have put into making this payout fair and reasonable. 

I do hope one day the tipjar issue get resolved.

**Comment by patricklodder** [comment]
Small note about the timing. I came back from vacay this weekend into a whirlwind of urgent issues, so I'll have to take a couple days more for the contribution analysis - apologies.

Please expect the full analysis here the coming weekend (of Aug 29).

**Comment by patricklodder** [comment]
Please find weighted contributions below for v1.14.6. Because some PRs were collaborative / backports and there are often multiple reviewers, there are 3 columns for authors and reviewers each, each with a column "c" that notes the weight of the contribution. Weights have been assigned per explanation in the OP.

You can check the completeness of PRs by doing:
```
git log --merges --pretty='"%h","%s"' v1.14.5..v1.14.6
```

And of course you can follow the linked PR in the table and request me to weigh something differently or you disagree, **just comment under this thread**. I have tried to make a short comment of "why" to each weight - so if you disagree please let me know why, or just ask me to explain further.

**Some fun stats before boring table:** we had 102 contributions, from 46 authors, of which only 13 were backports from Bitcoin Core. This time, we have had 24 reviewers, which is a great improvement over the past. Thanks everyone ❤️

| commit | pull request | author | c | author | c | author | c | reviewer | c | reviewer | c | reviewer | c | comment
| :--- | ---: | :--- | ---: | :--- | ---: | :--- | ---: | :--- | ---: | :--- | ---: | :--- | ---: | :--- 
| 3a29ba6d49 | #3048 | patricklodder | 1 |  |  |  |  | chromatic | 1 |  |  |  |  | maintainer task
| 6517a495fa | #3036 | rnohach | 20 | patricklodder | 0 |  |  | jimkov | 5 | ch0k1 | 5 | chromatic | 1 | major translation update
| ce770c4989 | #3047 | patricklodder | 0 |  |  |  |  | chromatic | 0 |  |  |  |  | credited under #3046
| 7d764249ef | #3029 | qlpqlp | 20 |  |  |  |  | justandrea1111 | 5 | patricklodder | 1 |  |  | major translation update
| 6add415c13 | #3050 | xanimo | 0 |  |  |  |  | patricklodder | 0 |  |  |  |  | key push only
| ad8625a031 | #3026 | patricklodder | 5 |  |  |  |  | maniixer | 1 | chromatic | 1 |  |  | translation update
| df20a6656a | #3032 | patricklodder | 5 |  |  |  |  | chromatic | 1 |  |  |  |  | new dev tool
| 0db6264e6c | #3049 | patricklodder | 0 |  |  |  |  | xanimo | 0 | chromatic | 0 |  |  | key push only
| 4cf5285ab0 | #3045 | patricklodder | 1 |  |  |  |  | chromatic | 1 |  |  |  |  | maintainer task
| d14cf7cf09 | #3031 | patricklodder | 1 |  |  |  |  | chromatic | 1 |  |  |  |  | maintainer task
| 4e23a29884 | #3042 | patricklodder | 5 |  |  |  |  | chromatic | 1 |  |  |  |  | major doc update
| b5ccb2fe00 | #3046 | chromatic | 1 |  |  |  |  | patricklodder | 0 | richdevx | 1 |  |  | trivial doc update
| 9296c93650 | #3044 | patricklodder | 0 |  |  |  |  | chromatic | 0 |  |  |  |  | credited under #3035
| c8b3a2f237 | #3043 | patricklodder | 0 |  |  |  |  | chromatic | 0 |  |  |  |  | credited under #3035
| 0cb4729a0c | #3038 | patricklodder | 1 |  |  |  |  | chromatic | 1 |  |  |  |  | maintainer task
| 96f9d2446d | #3037 | chromatic | 5 |  |  |  |  | xanimo | 1 | patricklodder | 1 |  |  | non-trivial bugfix
| 22884709d7 | #3008 | patricklodder | 5 | sipa | 5 | jonatack | 5 | xanimo | 1 | chromatic | 1 |  |  | non-trivial backport
| aa43aed7c2 | #3027 | patricklodder | 1 |  |  |  |  | chromatic | 1 |  |  |  |  | maintainer task
| 249a540dc2 | #3033 | patricklodder | 1 |  |  |  |  | chromatic | 0 |  |  |  |  | trivial doc fix
| 74633f7d5f | #3034 | patricklodder | 1 |  |  |  |  | chromatic | 0 |  |  |  |  | trivial doc fix
| dd6b6923fc | #3035 | patricklodder | 1 |  |  |  |  | chromatic | 1 | alamshafil | 1 |  |  | maintainer task
| 1729aaaf1f | #3030 | patricklodder | 5 | alamshafil | 5 | sdaftuar | 5 | chromatic | 1 |  |  |  |  | combination of multiple backports
| 82d1997235 | #3014 | mishaboar | 5 |  |  |  |  | patricklodder | 0 |  |  |  |  | translation update
| d5f4e810a4 | #2409 | patricklodder | 1 | fdoving | 5 | hebasto | 5 | chromatic | 1 | xanimo | 1 |  |  | dependency backport
| c6900bbb7b | #3022 | dogespacewizard | 5 |  |  |  |  | patricklodder | 0 |  |  |  |  | translation update
| 4674dc1b82 | #2910 | patricklodder | 1 |  |  |  |  | chromatic | 1 |  |  |  |  | trivial build fix
| c5ab255fbb | #3015 | chromatic | 5 |  |  |  |  | patricklodder | 1 |  |  |  |  | non-trivial bugfix
| d0bc620154 | #2964 | true-mining | 0 | matheusbach | 5 |  |  | patricklodder | 1 |  |  |  |  | fee plan implementation, credited to author
| 54a6dbe0bb | #2974 | chromatic | 20 |  |  |  |  | patricklodder | 1 |  |  |  |  | new feature
| 666e174f1f | #3011 | patricklodder | 1 |  |  |  |  | chromatic | 1 |  |  |  |  | maintainer task
| eecc9325bf | #3010 | patricklodder | 1 |  |  |  |  | chromatic | 1 |  |  |  |  | maintainer task
| 7c67087d61 | #3009 | patricklodder | 1 |  |  |  |  | chromatic | 1 |  |  |  |  | maintainer task
| b38a23cdfe | #3001 | patricklodder | 1 | thebluematt | 5 |  |  | chromatic | 1 |  |  |  |  | backport
| de26bbfbe1 | #2998 | patricklodder | 1 |  |  |  |  | chromatic | 1 |  |  |  |  | trivial text fixes
| 5121eef492 | #3003 | patricklodder | 5 |  |  |  |  | chromatic | 5 |  |  |  |  | qa fix, non-trivial review
| f91b29e1ca | #2997 | patricklodder | 5 |  |  |  |  | chromatic | 1 |  |  |  |  | non-trivial bugfix
| 77bb8b9f07 | #2995 | patricklodder | 1 |  |  |  |  | chromatic | 1 |  |  |  |  | trivial documentation
| d71725e699 | #2973 | patricklodder | 5 | micaelmalta | 5 | abcsxyz | 5 | chromatic | 1 | xanimo | 1 |  |  | also credits #2579 and #2396
| 96d34ae4da | #3000 | chromatic | 5 |  |  |  |  | patricklodder | 1 |  |  |  |  | non-trivial bugfix
| f132f56ac3 | #2996 | chromatic | 5 |  |  |  |  | patricklodder | 1 |  |  |  |  | non-trivial bugfix
| 5eb3a70cdc | #2988 | alamshafil | 20 |  |  |  |  | chromatic | 1 | patricklodder | 1 |  |  | new feature
| 132ebe6858 | #2994 | patricklodder | 1 |  |  |  |  | chromatic | 1 |  |  |  |  | trivial qa fix
| e22b54607c | #2991 | chromatic | 5 |  |  |  |  | patricklodder | 1 |  |  |  |  | non-trivial bugfix
| b3beca6cc4 | #2758 | patricklodder | 5 | gmaxwell | 5 |  |  | rnicoll | 1 |  |  |  |  | non-trivial backport
| 7c5901ec12 | #2936 | alamshafil | 5 |  |  |  |  | patricklodder | 5 | xanimo | 1 |  |  | dependency update
| 71507dde41 | #2934 | chromatic | 0 | michilumin | 20 |  |  | patricklodder | 1 | chromatic | 1 |  |  | enhancement, only review credited to PR author per request
| c6cd7efb56 | #2986 | patricklodder | 1 |  |  |  |  | chromatic | 1 |  |  |  |  | tiny qa improvement
| ed487e80c7 | #2976 | patricklodder | 5 |  |  |  |  | chromatic | 1 |  |  |  |  | new non-trivial test
| 99c19a1a34 | #2978 | patricklodder | 1 |  |  |  |  | chromatic | 0 |  |  |  |  | maintainer task
| 3c2a095361 | #2970 | h-mali | 5 |  |  |  |  | miladghanei | 1 | patricklodder | 0 |  |  | newly translated readme
| 1fcd30fa23 | #2798 | mvshishkov | 5 |  |  |  |  | abapchaoliu | 1 | patricklodder | 0 |  |  | significant readme translation update
| cf65be1dc1 | #2968 | scarletletters | 5 |  |  |  |  | matheusbach | 5 | patricklodder | 0 |  |  | newly translated readme
| 1718d3c467 | #2954 | patricklodder | 5 |  |  |  |  | chromatic | 1 |  |  |  |  | process change
| baf3e4cd72 | #2971 | vijaychowthri | 5 |  |  |  |  | santoshsivaji | 1 | patricklodder | 0 |  |  | translation update
| e4e8c4b48b | #2953 | jadehamel | 1 |  |  |  |  | chromatic | 0 | patricklodder | 0 | alamshafil | 0 | grammar fix
| 7b56cdbf14 | #2807 | steve-cw-chung | 10 |  |  |  |  | dogespacewizard | 1 |  |  |  |  | second half of major translation update, other half #2779
| 6fd30245b4 | #2956 | katesalazar | 1 |  |  |  |  | patricklodder | 0 |  |  |  |  | tiny patch
| 159786718f | #2912 | mishaboar | 20 |  |  |  |  | dogecoinitalia | 5 | michilumin | 1 |  |  | major translation update
| 23eb7b3cc0 | #2947 | maniixer | 20 |  |  |  |  | staartvos | 5 | patricklodder | 1 |  |  | major translation update
| 4d837b2395 | #2903 | thisiskeanyvy | 20 |  |  |  |  | martyornot | 5 | chromatic | 1 |  |  | major translation update
| bb5e476975 | #2698 | chromatic | 20 |  |  |  |  | patricklodder | 1 |  |  |  |  | new feature
| 10c294d2eb | #2937 | edtubbs | 5 |  |  |  |  | patricklodder | 1 |  |  |  |  | enhancement
| ecdba76ab3 | #2882 | patricklodder | 5 |  |  |  |  | edtubbs | 1 | chromatic | 1 |  |  | non-trivial enhancement
| 983ce7720f | #2909 | patricklodder | 1 | romanz | 5 |  |  | chromatic | 1 |  |  |  |  | backport
| c6a285828e | #2923 | alamshafil | 5 |  |  |  |  | patricklodder | 1 | chromatic | 1 |  |  | feature, non-trivial backport
| 0e02aea03a | #2870 | chromatic | 5 |  |  |  |  | patricklodder | 1 |  |  |  |  | process change
| 08d2d5173b | #2754 | sinetek | 5 |  |  |  |  | patricklodder | 1 | chromatic | 1 |  |  | enhancement
| 4008d95e53 | #2895 | motahharm | 1 |  |  |  |  | patricklodder | 0 | chromatic | 0 |  |  | trivial readme change
| 9c79c2de2f | #2860 | patricklodder | 0 |  |  |  |  | chromatic | 1 |  |  |  |  | credited with #2770
| d5b1719880 | #2869 | jadijadi | 5 |  |  |  |  | zeeamini | 1 | patricklodder | 1 |  |  | medium translation update
| f78b4b35e7 | #2907 | patricklodder | 0 |  |  |  |  | chromatic | 0 |  |  |  |  | reverted with #2882
| d483dfba80 | #2745 | vertiond | 5 |  |  |  |  | rnicoll | 1 | patricklodder | 1 |  |  | non-trivial bugfix
| 2a6421f79d | #2770 | just-an-dev | 5 |  |  |  |  | patricklodder | 1 |  |  |  |  | non-trivial backport
| cdeca5d581 | #2843 | patricklodder | 0 |  |  |  |  | rnicoll | 0 |  |  |  |  | credited with #2842
| 8fecf8d732 | #2842 | patricklodder | 5 |  |  |  |  | rnicoll | 1 |  |  |  |  | non-trivial enhancement
| 24a20ed523 | #2778 | chromatic | 1 | hebasto | 5 |  |  | fdoving | 1 | patricklodder | 1 |  |  | tiny backport
| 3024da3d6b | #2765 | rnicoll | 1 | jeremyrubin | 5 |  |  | patricklodder | 1 |  |  |  |  | tiny backport
| 646b1149dc | #2779 | dogespacewizard | 10 |  |  |  |  | steve-cw-chung | 1 | patricklodder | 0 |  |  | first half of major translation update, other half #2807
| 8b222213b3 | #2791 | iam-agf | 1 |  |  |  |  | rnicoll | 0 |  |  |  |  | generated images
| 94d7e452be | #2784 | bskrksyp9 | 1 |  |  |  |  | patricklodder | 0 |  |  |  |  | trivial doc fix
| a8026ba7d5 | #2768 | patricklodder | 5 |  |  |  |  | edtubbs | 1 | rnicoll | 1 |  |  | new feature
| da8dae3eb3 | #2756 | patricklodder | 1 | marcofalke | 5 |  |  | rnicoll | 1 |  |  |  |  | backport
| 190d48e487 | #2757 | patricklodder | 1 | kazcw | 5 |  |  | fdoving | 1 | rnicoll | 1 |  |  | backport
| e4e3371b6f | #2759 | patricklodder | 1 | ryanofsky | 5 |  |  | fdoving | 1 | sinetek | 1 | rnicoll | 1 | tiny backport
| b39c06aa90 | #2755 | patricklodder | 1 | practicalswift | 5 |  |  | fdoving | 1 | rnicoll | 1 |  |  | tiny backport
| 7019ead6a1 | #2687 | edtubbs | 20 |  |  |  |  | michilumin | 5 | patricklodder | 5 |  |  | new feature, non-trivial testing
| 2bc100ad10 | #2727 | patricklodder | 1 | laanwj | 5 |  |  | rnicoll | 1 |  |  |  |  | tiny backport
| 311b855a12 | #2637 | sinetek | 5 |  |  |  |  | patricklodder | 1 |  |  |  |  | non-trivial backport
| f6bbad704f | #2718 | likenttt | 1 |  |  |  |  | chromatic | 1 | patricklodder | 0 |  |  | trivial translation fix
| fd0798bb16 | #2716 | bmwant | 1 |  |  |  |  | patricklodder | 0 |  |  |  |  | trivial doc fix
| 89aeecc0dc | #2660 | patricklodder | 1 |  |  |  |  | edtubbs | 1 | rnicoll | 1 |  |  | trivial qa fix
| 4b3e6529e9 | #2693 | patricklodder | 5 |  |  |  |  | rnicoll | 1 |  |  |  |  | non-trivial bugfix
| 5449f441df | #2694 | patricklodder | 5 |  |  |  |  | rnicoll | 1 |  |  |  |  | new feature
| 543d896a49 | #2692 | ayushman17 | 1 |  |  |  |  | patricklodder | 0 |  |  |  |  | trivial doc fix
| 3b1376906d | #2680 | patricklodder | 1 |  |  |  |  | rnicoll | 0 |  |  |  |  | trivial build fix
| 23b4850130 | #2685 | patricklodder | 1 |  |  |  |  | rnicoll | 0 |  |  |  |  | trivial qa fix
| ca94ba2246 | #2683 | patricklodder | 1 |  |  |  |  | rnicoll | 0 |  |  |  |  | trivial qa fix
| 9102581176 | #2682 | patricklodder | 1 |  |  |  |  | rnicoll | 0 |  |  |  |  | trivial qa fix
| 8d6f5166f9 | #2684 | abcsxyz | 0 |  |  |  |  | patricklodder | 0 |  |  |  |  | key push only
| 93c94f1453 | #2678 | xanimo | 0 |  |  |  |  | patricklodder | 0 |  |  |  |  | key push only
| d52b659385 | #2677 | fdoving | 0 |  |  |  |  | patricklodder | 0 |  |  |  |  | key push only
| 4a5e4856c7 | #2679 | patricklodder | 1 |  |  |  |  | rnicoll | 1 |  |  |  |  | maintainer task
| | | TOTAL | 418 |  | 95 |  | 20 |  | 108 |  | 35 |  | 2 |

**Reply by patricklodder** [reply]
The one thing I have omitted from the table are security researchers because not all the security work has been disclosed yet. I will communicate with security people separately. The total weight of all security work is `56`. To not make this even more complicated, if there would be any challenge there I'll just fix that out of my own pocket.

**Comment by rnicoll** [comment]
Thanks! I've pulled that together into a spreadsheet, set up a weighted payout calculation from it. Do you have any specific thoughts on total payout pot size?

I think I've lost access to my old address spreadsheet, but will start collecting contributor addresses

**Reply by patricklodder** [reply]
I'm paying out all tips I received. My recommendation for the original tipjar would be to allocate 11% of whatever is in it at payout time for bigger releases from now on - that way, there is a future.

**Reply by patricklodder** [reply]
I'm sorry I made a mistake. I accidentally typed 10. I meant 11.

**Comment by chromatic** [comment]
I'm reviewing the categorization of all PRs. I have two minor quibbles, but I'm fine with the categorization/weighting as is:

- #3003 seems more than a trivial test fix; it took some work for me to understand it, so it was probably more than trivial to write it
- #2997 I think any core maintainer PR review for something that gets reverted should be 0; credit it to the PR that gets and stays merged

**Reply by patricklodder** [reply]
Re: #3003. I'll adjust that. The low-R trick was actually not so much work because I've done that before a couple of times but the analysis of what was causing the test to fail did take effort - it's hard to weigh one's own work so I try to err on the low end. I'll increase both author and reviewer weights on this.

Re: #2907 (I think you mean that one) - I actually missed it. I wouldn't make it a rule that reviewers of reverted PRs aren't credited though, because the review can be dissenting or causing the revert. In this case though, you are right. Will fix.

**Comment by michilumin** [comment]
so just inputting that I'm ok with this amount, mostly since I am definitely for making sure we motivate real help for upcoming releases. Specific ones that'll need the development muscle.  For those watching this discussion and reporting on it before the process is completed; I trust that since the reporting finger is now on the pulse of development, we won't hear that it's not happening, ever again. 

Also agree that generally weighting one's own work probably -should- err on the low side, but it's good to have input from more eyes that think that's not the case.

As for @chromatic 's comment on weighting (second) - yes, in this case; but reversions can happen for various reasons, and sometimes result in deferrals or still lead to solutions that do get utilized - or at least "could" - so it should be case by case.

**Comment by rnicoll** [comment]
Quick update - I've got addresses for most of the big recipients at least, Patrick I need yours (check your email please).

I've ended up travelling and left the spreadsheet on my desktop, so there'll be a delay until I'm back, but I'll raise transactions with the addresses I have, once I'm back, and we can include the list of users we don't have responses from in the announcement post.
